### ISSUE TITLE
Unlimited pngsave

### ISSUE THREAD

**Issue by jskrzypek** [libvips/libvips#2619]
We are using pyvips to write very long png files for use with a continuous roll printer, and are currently constrained by libpng's arbitrary 1M row limit. At high resolution the sizes we'd like to be able to write are in the 2-3 million pixel range.

#2419 introduced an unlimited flag for reading PNGs in pngload... could a similar approach of adding an unlimited flag work for writing PNGs in pngsave? 

I am not well-versed in C enough to know what to do to add this, but the libpng spec does seem to indicate that it's possible:

> User limits
> The PNG specification allows the width and height of an image to be as large as
> 2ˆ31-1 (0x7fffffff), or about 2.147 billion rows and columns. Since very few
> applications really need to process such large images, we have imposed an arbitrary
> 1-million limit on rows and columns. Larger images will be rejected immediately with
> a png_error() call. If you wish to override this limit, you can use
>     ```png_set_user_limits(png_ptr, width_max, height_max);```
> to set your own limits, or use ```width_max = height_max = 0x7fffffffL```
> to allow all valid dimensions (libpng may reject some very large images anyway because of potential buffer overflow conditions)

If you don't have time, perhaps you could give me some implementation pointers and I can see if I'm able to bash something together.

**Comment by jcupitt** [comment]
Hi @jskrzypek,

Wow, those are /huge/ PNGs. I've never tried making one that size (obviously). I didn't even know libpng had these soft limits.

As a quick workaround, you could try with libspng instead, the libpng replacement we've been working on:

https://github.com/libvips/libvips/pull/2536

You'll need to build libspng master, then rebuild that branch of libvips. I see:

```
$ vips black x.png 10 2000000
$ vipsheader x.png
x.png: 10x2000000 uchar, 1 band, b-w, pngload
```

So it seems to make huge PNGs without issue.

We should probably change the limits for libpng write too. libvips has a 100m pixel per axis limit by default, so maybe we should use that.

**Comment by jcupitt** [comment]
I've fixed 8.12 to raise this limit to 10m pixels per axis (the libvips default sdize limit).

**Comment by jskrzypek** [comment]
@jcupitt I forgot to come back and let you know, but this worked brilliantly! I created a long png file ~1.5m pixels with 0 errors, so I'll close this issue. Out of curiosity, do you have an idea of when 8.12.2 will be released? I'll build for linux too, I was just wondering.
